clash-lib/src/Clash/Driver/BrokenGhcs.hs

```diff
@@ -0,0 +1,156 @@
+-- | Current OS. Currently only recognizes Linux, Windows, and macOS.
+os :: OS
+os = case System.Info.os of
+  "darwin" -> OSDarwin
+  "linux" -> OSLinux
+  "mingw32" -> OSMinGW32
+  _ -> OSUnknown
+
+-- | What OS GHC is broken on (or all)
+data BrokenOn = All | SomeOs OS
+
+data GhcVersion
+  = Ghc Int Int Int
+  -- ^ @A.B.C@, for example @9.10.2@
+  deriving (Eq, Ord)
+
+data GhcRange = GhcRange
+  { from :: GhcVersion
+  -- ^ Start of range, inclusive
+  , to :: GhcVersion
+  -- ^ End of range, exclusive
+  }
+
+-- | Check if a 'GhcVersion' is within a 'GhcRange'
+ghcInRange :: GhcVersion -> GhcRange -> Bool
+ghcInRange ghc GhcRange{from, to} = from <= ghc && ghc < to
+
+-- | Construct a range of all GHC versions matching a major version
+ghcMajor :: Int -> Int -> GhcRange
+ghcMajor major minor = GhcRange{from=Ghc major minor 0, to=Ghc (major + 1) 0 0}
+
+data Why = Why
+  { what :: String
+    -- ^ What is broken
+  , solution :: String
+    -- ^ What can be done to work around or solve the issue
+  , issue :: String
+    -- ^ Link to issue
+  , brokenOn :: [(BrokenOn, GhcRange)]
+    -- ^ What operation systems are affected
+  }
+
+-- | Get current GHC version expressed as a triple. It probably does something
+-- non-sensible on unreleased GHCs.
+ghcVersion :: GhcVersion
+ghcVersion = Ghc major minor patch
+ where
+  (major, minor, patch) =
+    case fullCompilerVersion of
+      Version{versionBranch} ->
+        case versionBranch of
+          [] -> (0, 0, 1)
+          [a] -> (a, 0, 1)
+          [a, b] -> (a, b, 1)
+          [a, b, c] -> (a, b, c)
+          (a:b:c:_) -> (a, b, c)
+
+-- | Pretty print 'Why' into an error message
+whyPp :: Why -> String
+whyPp Why{what, solution, issue}= [I.i|
+  Clash has known issues on #{major}.#{minor}.#{patch} on your current
+  OS. While not completely preventing the compiler from working, we recommend
+  switching to another GHC version. Symptoms:
+
+    #{what}
+
+  Consider the following work around or solution:
+
+    #{solution}
+
+  More information can be found at:
+
+    #{issue}
+
+  If you want to ignore this message, pass the following flag to Clash:
+
+    -fclash-ignore-broken-ghcs
+  |]
+ where
+  Ghc major minor patch = ghcVersion
+
+-- | Which GHCs are broken and why
+brokenGhcs :: [Why]
+brokenGhcs = [brokenClashCores, brokenTypeErrors, slowStarts]
+ where
+  brokenClashCores = Why
+    { what = "GHC fails to compile clash-cores"
+    , solution = "Upgrade to GHC 9.4 or downgrade to GHC 8.10"
+    , issue = "<no link>"
+    , brokenOn = [(SomeOs OSMinGW32, ghcMajor 9 0)]
+    }
+
+  brokenTypeErrors = Why
+    { what = "Clash type error messages are indecipherable"
+    , solution = "Upgrade to GHC 9.4 or downgrade to GHC 9.0"
+    , issue = "<no link>"
+    , brokenOn = [(All, ghcMajor 9 2)]
+    }
+
+  slowStarts = Why
+    { what = "Clash starts really slowly from GHC 9.4.8 up to and including 9.6.2"
+    , solution = "Downgrade to GHC 9.4.7 or upgrade to GHC 9.6.3 or newer"
```

Comment:
```suggestion
    , solution = "Upgrade to GHC 9.6.3 or newer, or downgrade to GHC 9.4.7"
```

Ordered from most preferred to least (just like the previous entry).

Reply:
Done